code/hsec-sync/src/Security/Advisories/Sync.hs

```diff
@@ -1,76 +1,95 @@
 {-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE LambdaCase #-}
 
 module Security.Advisories.Sync
-  ( Repository (..),
-    defaultRepository,
+  ( Snapshot (..),
+    SnapshotUrl (..),
+    UpdatesUrl (..),
+    defaultSnapshot,
+    githubSnapshot,
     SyncStatus (..),
     sync,
     RepositoryStatus (..),
     status,
   )
 where
 
-import Data.Time (zonedTimeToUTC)
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Trans.Except (runExceptT, withExceptT)
 import Security.Advisories.Sync.Atom
-import Security.Advisories.Sync.Git
+import Security.Advisories.Sync.Snapshot
+import Security.Advisories.Sync.Url
+
+data Snapshot = Snapshot
+  { snapshotRoot :: FilePath,
+    snapshotUrl :: SnapshotUrl,
+    updatesUrl :: UpdatesUrl
+  }
+
+defaultSnapshot :: Snapshot
+defaultSnapshot =
+  githubSnapshot
+    "security-advisories"
+    "https://github.com/haskell/security-advisories"
+    "generated/snapshot-export"
+
+githubSnapshot :: FilePath -> String -> String -> Snapshot
+githubSnapshot root repoUrl repoBranch =
+  Snapshot
+    { snapshotRoot = root,
+      snapshotUrl = SnapshotUrl $ ensureFile (mkUrl [repoUrl, "archive/refs/heads", repoBranch]) <> ".tar.gz",
```

Comment:
There's an interesting consequence of using the tarballs as generated by GitHub:
```
% curl --silent -L https://github.com/frasertweedale/security-advisories/archive/refs/heads/generated/snapshot-export.tar.gz  | tar -tz |head -n 5
security-advisories-generated-snapshot-export/
security-advisories-generated-snapshot-export/advisories/
security-advisories-generated-snapshot-export/advisories/hackage/
security-advisories-generated-snapshot-export/advisories/hackage/aeson/
security-advisories-generated-snapshot-export/advisories/hackage/aeson/HSEC-2023-0001.md
```

GitHub automatically adds a directory path.  Because `sync` extracts the tarball as-is, subsequent commands using the same `--snapshot-root` will get confused:

```
% hsec-sync sync -d ROOT -r https://github.com/frasertweedale/security-advisories
Snapshot at "ROOT" from "https://github.com/frasertweedale/security-advisories/archive/refs/heads/generated/snapshot-export.tar.gz"
Snapshot just created

% hsec-sync status -d ROOT -r https://github.com/frasertweedale/security-advisories
Directory is incoherent

% ll ROOT
total 0K
drwxr-xr-x. 1 ftweedal ftweedal 46 Jun 15 17:56 security-advisories-generated-snapshot-export

% ll ROOT/security-advisories-generated-snapshot-export 
total 4K
drwxr-xr-x. 1 ftweedal ftweedal 30 Jun 15 17:56 advisories
-rw-r--r--. 1 ftweedal ftweedal 67 Jun 14 16:15 snapshot.json
```

The leading path component is also dependent on the repo name and branch name. 

One approach might be to remove the leading path component during extraction.  I haven't
tried it yet, but probably something like this:

```haskell
Tar.unpack dir . magic . Tar.read =<< BL.readFile tar
  where
    magic = Tar.foldEntries f Tar.Done Tar.Fail
    f entry z
      | entryIsTheTopLevelDir = z  -- omit it
      | otherwise = Tar.Next (stripEntry entry) z
  stripEntry entry = stuff happens
```